src/dbnode/storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -180,14 +180,40 @@ func (s *peersSource) Read(
 			continue
 		}
 
-		r, err := s.readIndex(md,
-			namespace.IndexRunOptions.ShardTimeRanges,
-			instrCtx.span,
-			cache,
-			namespace.IndexRunOptions.RunOptions,
+		var (
+			opts           = namespace.IndexRunOptions.RunOptions
+			indexBlockSize = md.Options().IndexOptions().BlockSize()
+			idxOpts        = md.Options().IndexOptions()
+			r              result.IndexBootstrapResult
+			err            error
 		)
-		if err != nil {
-			return bootstrap.NamespaceResults{}, err
+		if s.shouldPersist(opts) {
+			// Only attempt to bootstrap index if we've persisted tsdb data.
+			r, err = s.readIndex(md,
+				namespace.IndexRunOptions.ShardTimeRanges,
+				instrCtx.span,
+				cache,
+				opts,
+			)
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+		} else {
+			// Mark index ranges as fulfilled if we did not persist any tsdb data (e.g. snapshot data).
+			r = result.NewIndexBootstrapResult()
+			min, max := namespace.IndexRunOptions.ShardTimeRanges.MinMax()
+			for start := min.Truncate(indexBlockSize); start.Before(max); start = start.Add(indexBlockSize) {
+				shardTimeRanges := result.NewShardTimeRangesFromRange(start, start.Add(indexBlockSize), namespace.Shards...)
+				if err := r.IndexResults().MarkFulfilled(
+					start,
+					shardTimeRanges,
+					// NB(bodu): By default, we always load bootstrapped data into the default index volume.
+					idxpersist.DefaultIndexVolumeType,
+					idxOpts,
```

Comment:
I see there are segment arrays inside `IndexBlock` struct inside `IndexResults`. I suppose `MarkFulfilled()` only sets time ranges, but do not append any segments, right?

I wonder if this could cause any unexpected behaviour, since I see there are some logic in `index/block.go` that takes into account fulfilled ranges and uses segments from index results:

https://github.com/m3db/m3/blob/812d585fddc53d7a533d69bde777efbfbc7bf29e/src/dbnode/storage/index/block.go#L1015-L1041

Reply:
Yea we had a conversation about this in slack as well. From my POV, it seems like we should only be peers bootstraping shard time ranges that we don't already have on the node. Newly added shards that are `INITIALIZING`. 

In the case that a shard has been moved and moved back I think the fs bootstrapper should bootstrap the older shard time ranges during which the node owned the shard and the peers bootstrapper will do the rest. 

Also, w.r.t. the series that were transparently indexed during loading of mutable data blocks (snapshots) - bootstrapped index segments live inside of `shardRangesSegmentsByVolumeType` within the index block while writes go to either the `mutableSegment` or the latest `coldMutableSegments` (in this case the `mutableSegment` since we are bootstrapping the warm time range.